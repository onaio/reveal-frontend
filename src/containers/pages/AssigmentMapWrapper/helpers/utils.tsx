import { Dictionary } from '@onaio/utils';
import { Feature } from '@turf/turf';
import { History } from 'history';
import { EventData, Map } from 'mapbox-gl';
import { AssignmentMapWrapperProps } from '..';
import { PLAN_TYPES_WITH_MULTI_JURISDICTIONS } from '../../../../configs/env';
import { getPlanType } from '../../../../helpers/utils';
import { SELECTION_REASON } from '../../../../store/ducks/opensrp/hierarchies/constants';
import { TreeNode } from '../../../../store/ducks/opensrp/hierarchies/types';
import { nodeIsSelected } from '../../../../store/ducks/opensrp/hierarchies/utils';

/**
 * Handler to get current parent id from clicked jurisdiction on map
 * @param props - AssignmentMapWrapper component props
 * @param setMapParent - function to set map parent id to state
 * @param history - history session object
 */

export const onJurisdictionClick = (
  props: AssignmentMapWrapperProps,
  setMapParent: (currentId: string) => void,
  history: History
) => {
  function handleJurisdictionClick(_: Map, e: EventData) {
    // Destructure event data
    const { point, target, originalEvent } = e;
    const {
      currentParentNode,
      rootJurisdictionId,
      selectNodeCreator,
      deselectNodeCreator,
      currentChildren,
      plan,
      baseAssignmentURL,
    } = props;
    // grab underlying features from map
    const features: Feature[] = target.queryRenderedFeatures(point);
    if (!features.length) {
      return false;
    }

    for (const feature of features) {
      const activeJurisdictionId =
        (feature && feature.id) ||
        (feature && feature.properties && feature.properties.jurisdiction_id);
      if (activeJurisdictionId) {
        const currentId = activeJurisdictionId;
        let activeCurrentNode: TreeNode | undefined;

        if (currentParentNode) {
          // ensure currentparentnode is not undefined first
          // Handle selection for admin level 0
          if (
            currentParentNode.model.id === rootJurisdictionId &&
            currentId === currentParentNode.model.id
          ) {
            activeCurrentNode = currentParentNode;
          } else {
            // Handle selection for admin level > 0
            activeCurrentNode = currentChildren.find(
              (node: TreeNode) => node.model.id === currentId
            );
          }
        }
        const thePlanType = getPlanType(plan);
        if (
          activeCurrentNode &&
          originalEvent.altKey &&
          PLAN_TYPES_WITH_MULTI_JURISDICTIONS.includes(thePlanType)
        ) {
          if (!nodeIsSelected(activeCurrentNode)) {
            selectNodeCreator(
              rootJurisdictionId,
              activeCurrentNode.model.id,
              plan.identifier,
              SELECTION_REASON.USER_CHANGE
            );
          } else {
            deselectNodeCreator(
              rootJurisdictionId,
              activeCurrentNode.model.id,
              plan.identifier,
              SELECTION_REASON.USER_CHANGE
            );
          }
          setMapParent(currentId);
        } else {
          if (activeCurrentNode && activeCurrentNode.hasChildren()) {
            history.push(`${baseAssignmentURL}/${currentId}`);
          }
        }
        break;
      }
    }
  }
  return handleJurisdictionClick;
};

/**
 * Map component popup handler
 * @param map - mapbox map instance
 * @param event - map event object
 */

export const buildMouseMoveHandler = (map: Map, event: EventData) => {
  const popup = new mapboxgl.Popup({
    closeButton: false,
  });
  // Added this part to remove duplicate popup containers showing on map when hovering
  const nodes: any = document.getElementsByClassName('mapboxgl-popup');
  let node: HTMLDivElement;
  if (nodes && nodes.length) {
    for (node of nodes) {
      if (node.style) {
        node.style.display = 'none';
      }
    }
  }
  let content: string = '';
  let feature: Feature | any = {};
  popup.remove();
  // grab underlying features from map
  const features: any = event.target.queryRenderedFeatures(event.point) as Dictionary[];
  //
  if (!features.length) {
    return;
  }

  // Change the cursor style to pointer if features exist.
  map.getCanvas().style.cursor = features.length ? 'pointer' : 'grab';

  // loop through features, setting the name property to content var
  for (feature of features) {
    if (feature && feature.properties && feature.properties.name) {
      content += `<div class="jurisdiction-name"><center>${feature.properties.name}</center></div>`;
    }
    /**
     * Break out of loop once content is set
     * This helps in a case where we could have 2 or more identical features
     */
    if (content) {
      break;
    }
  }
  /**
   * Add popup to map after content is set
   */
  if (content.length) {
    popup
      .setLngLat(map.unproject(event.point))
      .setHTML(content)
      .addTo(map);
  } else {
    // clear existing popups if content is empty
    popup.remove();
    return;
  }
  return true;
};
