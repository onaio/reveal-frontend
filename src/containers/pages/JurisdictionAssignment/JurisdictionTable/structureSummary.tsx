/** gives a summary of the selected jurisdictions as well as their structures count */
import ListView from '@onaio/list-view';
import reducerRegistry from '@onaio/redux-reducer-registry';
import React from 'react';
import { connect } from 'react-redux';
import { useParams } from 'react-router';
import Button from 'reactstrap/lib/Button';
import { Store } from 'redux';
import HeaderBreadcrumb, {
  Page,
} from '../../../../components/page/HeaderBreadcrumb/HeaderBreadcrumb';
import {
  CONTINUE_TO_NEXT_STEP,
  NAME,
  NO_ROWS_FOUND,
  SELECTED_JURISDICTIONS,
  STRUCTURES_COUNT,
} from '../../../../configs/lang';
import { AUTO_ASSIGN_JURISDICTIONS_URL } from '../../../../constants';
import hierarchyReducer, {
  Filters,
  getNodesInSelectedTree,
  getSelectedHierarchy,
  reducerName as hierarchyReducerName,
} from '../../../../store/ducks/opensrp/hierarchies';
import { TreeNode } from '../../../../store/ducks/opensrp/hierarchies/types';
import { NodeCell } from '../JurisdictionCell';

reducerRegistry.register(hierarchyReducerName, hierarchyReducer);

/** props for the Jurisdiction selector table component */
export interface JurisdictionSelectorTableProps {
  tree: TreeNode | undefined;
  currentParentId: string | undefined;
  rootJurisdictionId: string;
  currentParentNode: TreeNode | undefined;
  currentChildren: TreeNode[];
  onClickNext: any;
}

const defaultProps = {
  currentChildren: [],
  currentParentId: undefined,
  currentParentNode: undefined,
  onClickNext: () => {
    return;
  },
  rootJurisdictionId: '',
  tree: undefined,
};

/** This component should provide a summary of the structures  */
const SelectedStructuresTable = (props: JurisdictionSelectorTableProps) => {
  const { rootJurisdictionId, currentParentNode, currentChildren } = props;

  const { planId } = useParams();
  const baseUrl = `${AUTO_ASSIGN_JURISDICTIONS_URL}/${planId}/${rootJurisdictionId}`;

  // we will use the currentParentId prop to know if the parent node has
  // been set, if drilling down has begun.
  let derivedParentNode = currentParentNode;
  let derivedChildrenNodes = currentChildren;
  if (!props.currentParentId) {
    derivedParentNode = undefined;
    derivedChildrenNodes = currentParentNode ? [currentParentNode] : [];
  }

  /** creating breadCrumb props */
  let currentPage: Page = {
    label: '....',
    url: baseUrl,
  };
  const pages: Page[] = [];

  if (derivedParentNode) {
    const path = derivedParentNode.getPath();
    const lastNode = path.pop();

    pages.push(currentPage);

    path.forEach(nd => {
      pages.push({
        label: nd.model.label,
        url: `${baseUrl}/${nd.model.id}`,
      });
    });

    currentPage = {
      label: (lastNode as TreeNode).model.label,
    };
  }

  const breadCrumbProps = {
    currentPage,
    pages,
  };
  const data = derivedChildrenNodes.map(node => {
    return [
      <NodeCell key={`${node.model.id}-structures-summary`} node={node} baseUrl={baseUrl} />,
      node.model.meta.metaStructureCount,
    ];
  });
  const headerItems = [NAME, STRUCTURES_COUNT, SELECTED_JURISDICTIONS];
  const tableClass = 'table table-bordered';

  const listViewProps = {
    data,
    headerItems,
    tableClass,
  };

  return (
    <div>
      <HeaderBreadcrumb {...breadCrumbProps} />
      <ListView {...listViewProps} />
      {!data.length && (
        <div style={{ textAlign: 'center' }}>
          {NO_ROWS_FOUND}
          <hr />
        </div>
      )}
      {!!data.length && (
        <>
          <hr />
          <Button className="btn btn-success float-right mt-3" onClick={props.onClickNext}>
            {CONTINUE_TO_NEXT_STEP}
          </Button>
        </>
      )}
    </div>
  );
};

SelectedStructuresTable.defaultProps = defaultProps;

export { SelectedStructuresTable as JurisdictionTable };

/** map state to props interface  */
type MapStateToProps = Pick<
  JurisdictionSelectorTableProps,
  'currentChildren' | 'currentParentNode' | 'tree'
>;

const treeSelector = getSelectedHierarchy();
const nodeSelector = getNodesInSelectedTree();

/** maps props to store state */
const mapStateToProps = (
  state: Partial<Store>,
  ownProps: JurisdictionSelectorTableProps
): MapStateToProps => {
  const filters: Filters = {
    currentParentId: ownProps.currentParentId,
    rootJurisdictionId: ownProps.rootJurisdictionId,
  };
  const tree = treeSelector(state, filters);
  const nodes = nodeSelector(state, filters);
  return {
    currentChildren: nodes.childrenNodes,
    currentParentNode: nodes.parentNode,
    tree,
  };
};

export const ConnectedSelectedStructuresTable = connect(mapStateToProps)(SelectedStructuresTable);
