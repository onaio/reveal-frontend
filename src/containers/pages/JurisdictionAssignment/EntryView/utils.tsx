/** collection of hooks that are used in the jurisdiction Assignment pages
 * these hooks majorly focus on abstracting the opensrp service calls.
 */
import { Result } from '@onaio/utils/dist/types/types';
import React from 'react';
import { ActionCreator } from 'redux';
import { getAncestors } from '../../../../components/TreeWalker/helpers';
import {
  COULD_NOT_LOAD_JURISDICTION,
  COULD_NOT_LOAD_JURISDICTION_HIERARCHY,
  COULD_NOT_LOAD_PLAN,
} from '../../../../configs/lang';
import { PlanDefinition } from '../../../../configs/settings';
import {
  loadJurisdiction,
  LoadOpenSRPHierarchy,
} from '../../../../helpers/dataLoading/jurisdictions';
import { loadOpenSRPPlan } from '../../../../helpers/dataLoading/plans';
import { StartLoading, StopLoading } from '../../../../helpers/useLoadingReducer';
import { OpenSRPService } from '../../../../services/opensrp';
import { FetchedTreeAction } from '../../../../store/ducks/opensrp/hierarchies';
import { RawOpenSRPHierarchy, TreeNode } from '../../../../store/ducks/opensrp/hierarchies/types';
import { AddPlanDefinitionAction } from '../../../../store/ducks/opensrp/PlanDefinition';
import { HandleBrokenPage } from '../helpers/utils';

/** hook to load a plan and dispatch to store
 * @param routePlanId - id of the plan, usually got from the route
 * @param plan - if the plan exists, otherwise null
 * @param serviceClass -  opensrp service
 * @param fetchPlanCreator - action creator to be called with plan payload once fetched
 * @param handleBrokenPage - helper to handle errors
 * @param stopLoading- helper to denote when loading should stop
 * @param startLoading - helper to denote when loading should start
 */
export function usePlanEffect(
  routePlanId: string,
  plan: PlanDefinition | null,
  serviceClass: typeof OpenSRPService,
  fetchPlanCreator: ActionCreator<AddPlanDefinitionAction>,
  handleBrokenPage: HandleBrokenPage,
  stopLoading: StopLoading,
  startLoading: StartLoading
) {
  React.useEffect(() => {
    // create promise to fetch the plan
    const planLoadingKey = 'planMetadata';
    startLoading(planLoadingKey, !plan);
    const planDataPromise = loadOpenSRPPlan(routePlanId, serviceClass, fetchPlanCreator).catch(
      _ => {
        handleBrokenPage(COULD_NOT_LOAD_PLAN);
      }
    );

    Promise.all([planDataPromise])
      .finally(() => {
        stopLoading(planLoadingKey);
      })
      .catch(e => handleBrokenPage(e.message));
  }, []);
}

/** gets the rootJurisdictionId - uses any of the assigned/existing jurisdictions
 * @param plan - the plan if it exists, otherwise null
 * @param serviceClass - the opensrpService
 * @param handleBrokenPage - helper to handle errors
 * @param stopLoading - helper to stop loading
 * @param startLoading - helper to start loading
 * @param tree - the tree if it exists, used to know if we should set loading to true.
 */
export function useGetRootJurisdictionId(
  plan: PlanDefinition | null,
  serviceClass: typeof OpenSRPService,
  handleBrokenPage: HandleBrokenPage,
  stopLoading: StopLoading,
  startLoading: StartLoading,
  tree?: TreeNode
) {
  const [rootJurisdictionId, setRootJurisdictionId] = React.useState<string>('');
  React.useEffect(() => {
    if (plan) {
      const oneOfJurisdictions = plan.jurisdiction.map(
        jurisdictionCode => jurisdictionCode.code
      )[0];
      // get the rootJurisdiction
      const rootJurisdictionKey = 'rootJurisdiction';
      startLoading(rootJurisdictionKey, !tree);
      loadJurisdiction(oneOfJurisdictions, serviceClass)
        .then(result => {
          if (!result || result.error) {
            stopLoading(rootJurisdictionKey);
            throw new Error(COULD_NOT_LOAD_JURISDICTION);
          }
          if (result.value) {
            // TODO: review this - we already have the entire hierarchy in
            // the hierarchy reducer module, do we need to always call OpenSRP?
            // -> this is where we add the hierarchy. we are getting a single jurisdiction from the plan
            // using that to get the root jurisdiction, and then requesting the hierarchy to save to store.
            getAncestors(result.value)
              .then(ancestors => {
                if (ancestors.value) {
                  // get the first ancestor
                  const rootJurisdiction = ancestors.value[0];
                  setRootJurisdictionId(rootJurisdiction.id);
                  stopLoading(rootJurisdictionKey);
                } else {
                  throw new Error(COULD_NOT_LOAD_JURISDICTION);
                }
              })
              .catch(e => {
                return e;
              });
          }
        })
        .catch(error => {
          handleBrokenPage(error.message);
        });
    }
  }, [plan]);

  return rootJurisdictionId;
}

/** get the jurisdiction Tree given the rootJurisdiction Id
 * @param rootJurisdictionId - id of top level jurisdiction
 * @param startLoading - helper to start loading
 * @param treeFetchedCreator - action creator to be called with tree payload once fetched
 * @param stopLoading - helper to stop loading
 * @param handleBrokenPage - helper to handle page errors
 * @param tree - the tree if it exists, used to know if we should set loading to true.
 */
export function useGetJurisdictionTree(
  rootJurisdictionId: string,
  startLoading: StartLoading,
  treeFetchedCreator: ActionCreator<FetchedTreeAction>,
  stopLoading: StopLoading,
  handleBrokenPage: HandleBrokenPage,
  tree?: TreeNode
) {
  React.useEffect(() => {
    const params = {
      return_structure_count: true,
    };
    if (rootJurisdictionId) {
      startLoading(rootJurisdictionId, !tree);
      LoadOpenSRPHierarchy(rootJurisdictionId, OpenSRPService, params)
        .then((apiResponse: Result<RawOpenSRPHierarchy>) => {
          if (apiResponse.value) {
            const responseData = apiResponse.value;
            treeFetchedCreator(responseData);
            stopLoading(rootJurisdictionId);
          }
          if (apiResponse.error) {
            throw new Error(COULD_NOT_LOAD_JURISDICTION_HIERARCHY);
          }
        })
        .catch(() => {
          handleBrokenPage(COULD_NOT_LOAD_JURISDICTION_HIERARCHY);
          stopLoading(rootJurisdictionId);
        });
    }
  }, [rootJurisdictionId]);
}
