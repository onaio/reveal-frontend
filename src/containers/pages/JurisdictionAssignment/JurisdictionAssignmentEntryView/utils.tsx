import { Result } from '@onaio/utils/dist/types/types';
import React from 'react';
import { ActionCreator } from 'redux';
import { getAncestors } from '../../../../components/TreeWalker/helpers';
import {
  COULD_NOT_LOAD_JURISDICTION,
  COULD_NOT_LOAD_JURISDICTION_HIERARCHY,
  COULD_NOT_LOAD_PLAN,
} from '../../../../configs/lang';
import { PlanDefinition } from '../../../../configs/settings';
import {
  loadJurisdiction,
  LoadOpenSRPHierarchy,
} from '../../../../helpers/dataLoading/jurisdictions';
import { loadOpenSRPPlan } from '../../../../helpers/dataLoading/plans';
import { OpenSRPService } from '../../../../services/opensrp';
import { RawOpenSRPHierarchy, TreeNode } from '../../../../store/ducks/opensrp/hierarchies/types';
import { AddPlanDefinitionAction } from '../../../../store/ducks/opensrp/PlanDefinition';

export function usePlanEffect(
  routePlanId: string,
  plan: PlanDefinition | null,
  serviceClass: typeof OpenSRPService,
  fetchPlanCreator: ActionCreator<AddPlanDefinitionAction>,
  handleBrokenPage: any,
  stopLoading: any,
  startLoading: any
) {
  React.useEffect(() => {
    // create promise to fetch the plan
    const planAndMetaLoadingKey = 'planMetadata';
    startLoading(planAndMetaLoadingKey, !plan);
    const planDataPromise = loadOpenSRPPlan(routePlanId, serviceClass, fetchPlanCreator).catch(
      _ => {
        handleBrokenPage(COULD_NOT_LOAD_PLAN);
      }
    );

    Promise.all([planDataPromise])
      .finally(() => {
        stopLoading(planAndMetaLoadingKey);
      })
      .catch(e => handleBrokenPage(e.message));
  }, []);
}

export function useGetRootJurisdictionId(
  plan: PlanDefinition | null,
  serviceClass: typeof OpenSRPService,
  handleBrokenPage: any,
  stopLoading: any,
  startLoading: any,
  tree?: TreeNode
) {
  const [rootJurisdictionId, setRootJurisdictionId] = React.useState<string>('');
  React.useEffect(() => {
    if (plan) {
      const oneOfJurisdictions = plan.jurisdiction.map(
        jurisdictionCode => jurisdictionCode.code
      )[0];
      // get the rootJurisdiction
      const rootJurisdictionKey = 'rootJurisdiction';
      startLoading(rootJurisdictionKey, !tree);
      loadJurisdiction(oneOfJurisdictions, serviceClass)
        .then(result => {
          if (!result || result.error) {
            stopLoading(rootJurisdictionKey);
            throw new Error(COULD_NOT_LOAD_JURISDICTION);
          }
          if (result.value) {
            // TODO: review this - we already have the entire hierarchy in
            // the hierarchy reducer module, do we need to always call OpenSRP?
            // -> this is where we add the hierarchy. we are getting a single jurisdiction from the plan
            // using that to get the root jurisdiction, and then requesting the hierarchy to save to store.
            getAncestors(result.value)
              .then(ancestors => {
                if (ancestors.value) {
                  // get the first ancestor
                  const rootJurisdiction = ancestors.value[0];
                  setRootJurisdictionId(rootJurisdiction.id);
                  stopLoading(rootJurisdictionKey);
                } else {
                  throw new Error(COULD_NOT_LOAD_JURISDICTION);
                }
              })
              .catch(e => {
                return e;
              });
          }
        })
        .catch(error => {
          handleBrokenPage(error.message);
        });
    }
  }, [plan]);

  return rootJurisdictionId;
}

export function useGetJurisdictionTree(
  rootJurisdictionId: string,
  startLoading: any,
  treeFetchedCreator: any,
  stopLoading: any,
  handleBrokenPage: any,
  tree?: TreeNode
) {
  React.useEffect(() => {
    const params = {
      return_structure_count: true,
    };
    if (rootJurisdictionId) {
      startLoading(rootJurisdictionId, !tree);
      LoadOpenSRPHierarchy(rootJurisdictionId, OpenSRPService, params)
        .then((apiResponse: Result<RawOpenSRPHierarchy>) => {
          if (apiResponse.value) {
            const responseData = apiResponse.value;
            treeFetchedCreator(responseData);
            stopLoading(rootJurisdictionId);
          }
          if (apiResponse.error) {
            throw new Error(COULD_NOT_LOAD_JURISDICTION_HIERARCHY);
          }
        })
        .catch(() => {
          handleBrokenPage(COULD_NOT_LOAD_JURISDICTION_HIERARCHY);
          stopLoading(rootJurisdictionId);
        });
    }
  }, [rootJurisdictionId]);
}
