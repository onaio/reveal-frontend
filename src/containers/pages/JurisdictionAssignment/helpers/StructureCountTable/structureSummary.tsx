/** gives a summary of the selected jurisdictions as well as their structures count */
import ElementMap from '@onaio/element-map';
import ListView, { renderRowsFunc, renderRowsFuncType } from '@onaio/list-view';
import reducerRegistry from '@onaio/redux-reducer-registry';
import React from 'react';
import { connect } from 'react-redux';
import Button from 'reactstrap/lib/Button';
import { Store } from 'redux';
import HeaderBreadcrumb, {
  Page,
} from '../../../../../components/page/HeaderBreadcrumb/HeaderBreadcrumb';
import {
  CONTINUE_TO_NEXT_STEP,
  NAME,
  NO_ROWS_FOUND,
  STRUCTURES_COUNT,
  TOTAL,
} from '../../../../../configs/lang';
import { AUTO_ASSIGN_JURISDICTIONS_URL } from '../../../../../constants';
import hierarchyReducer, {
  Filters,
  getNodesInSelectedTree,
  reducerName as hierarchyReducerName,
} from '../../../../../store/ducks/opensrp/hierarchies';
import { TreeNode } from '../../../../../store/ducks/opensrp/hierarchies/types';
import { getNodeStructureCount } from '../../../../../store/ducks/opensrp/hierarchies/utils';
import { NodeCell } from '../JurisdictionCell';

reducerRegistry.register(hierarchyReducerName, hierarchyReducer);

/** props for the Jurisdiction selector table component */
export interface JurisdictionSelectorTableProps {
  planId: string;
  currentParentId?: string;
  rootJurisdictionId: string;
  currentParentNode?: TreeNode;
  currentChildren: TreeNode[];
  onClickNext: () => void;
}

const defaultProps = {
  currentChildren: [],
  onClickNext: () => {
    return;
  },
  rootJurisdictionId: '',
};

/** This component should provide a summary of the structures  */
const SelectedStructuresTable = (props: JurisdictionSelectorTableProps) => {
  const { rootJurisdictionId, currentParentNode, currentChildren, planId } = props;
  const baseUrl = `${AUTO_ASSIGN_JURISDICTIONS_URL}/${planId}/${rootJurisdictionId}`;

  // we will use the currentParentId prop to know if the parent node has
  // been set, if drilling down has begun.
  let derivedParentNode = currentParentNode;
  let derivedChildrenNodes = currentChildren;
  if (!props.currentParentId) {
    derivedParentNode = undefined;
    derivedChildrenNodes = currentParentNode ? [currentParentNode] : [];
  }

  /** creating breadCrumb props */
  let currentPage: Page = {
    label: '....',
    url: baseUrl,
  };
  const pages: Page[] = [];

  if (derivedParentNode) {
    const path = derivedParentNode.getPath();
    const lastNode = path.pop();

    pages.push(currentPage);

    path.forEach(nd => {
      pages.push({
        label: nd.model.label,
        url: `${baseUrl}/${nd.model.id}`,
      });
    });

    currentPage = {
      label: (lastNode as TreeNode).model.label,
    };
  }

  const breadCrumbProps = {
    currentPage,
    pages,
  };
  const data = derivedChildrenNodes.map(node => {
    return [
      <NodeCell key={`${node.model.id}-structures-summary`} node={node} baseUrl={baseUrl} />,
      getNodeStructureCount(node),
    ];
  });
  const headerItems = [NAME, STRUCTURES_COUNT];
  const tableClass = 'table table-bordered';

  /** Adds a row in the footer that contains the total number of structures of all
   * jurisdictions that are descendants of the current parent node
   */
  const renderRows: renderRowsFuncType = (rowData, tbClass, tdClass, trClass) => {
    const parentNodeStructureCount = currentParentNode
      ? getNodeStructureCount(currentParentNode)
      : 0;
    const items = [TOTAL, parentNodeStructureCount];
    const tfootRow = (
      <tr className={trClass}>
        <ElementMap items={items} HTMLTag="td" className={`${tdClass} text-bold`} />
      </tr>
    );
    return (
      <>
        {renderRowsFunc(rowData, tbClass, tdClass, trClass)}
        {parentNodeStructureCount !== 0 && <tfoot>{tfootRow}</tfoot>}
      </>
    );
  };

  const listViewProps = {
    data,
    headerItems,
    renderRows,
    tableClass,
  };

  return (
    <div>
      <HeaderBreadcrumb {...breadCrumbProps} />
      <ListView {...listViewProps} />
      {!data.length && (
        <div style={{ textAlign: 'center' }}>
          {NO_ROWS_FOUND}
          <hr />
        </div>
      )}
      <>
        <hr />
        <Button className="btn btn-success float-right mt-3" onClick={props.onClickNext}>
          {CONTINUE_TO_NEXT_STEP}
        </Button>
      </>
    </div>
  );
};

SelectedStructuresTable.defaultProps = defaultProps;

export { SelectedStructuresTable as JurisdictionTable };

/** map state to props interface  */
type MapStateToProps = Pick<
  JurisdictionSelectorTableProps,
  'currentChildren' | 'currentParentNode'
>;

const nodeSelector = getNodesInSelectedTree();

/** maps props to store state */
const mapStateToProps = (
  state: Partial<Store>,
  ownProps: JurisdictionSelectorTableProps
): MapStateToProps => {
  const filters: Filters = {
    currentParentId: ownProps.currentParentId,
    planId: ownProps.planId,
    rootJurisdictionId: ownProps.rootJurisdictionId,
  };
  const nodes = nodeSelector(state, filters);
  return {
    currentChildren: nodes.childrenNodes,
    currentParentNode: nodes.parentNode,
  };
};

export const ConnectedSelectedStructuresTable = connect(mapStateToProps)(SelectedStructuresTable);
