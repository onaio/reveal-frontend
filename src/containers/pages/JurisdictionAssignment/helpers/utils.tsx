import { useState } from 'react';
import { displayError } from '../../../../helpers/errors';
import { TreeNode } from '../../../../store/ducks/opensrp/hierarchies/types';
import { nodeIsSelected } from '../../../../store/ducks/opensrp/hierarchies/utils';

/** tells whether we should select the parent checkbox in the table header
 * @param currentParentNode - the current parent node
 * @param currentChildren - current children
 */
export const checkParentCheckbox = (
  currentParentNode: TreeNode | undefined,
  currentChildren: TreeNode[]
) => {
  let selected = true;
  if (currentParentNode) {
    selected = selected && nodeIsSelected(currentParentNode);
  } else {
    // when currentParent node is undefined but there are currentChildren usually at the top level
    if (currentChildren.length > 0) {
      currentChildren.forEach(node => {
        selected = selected && nodeIsSelected(node);
      });
    } else {
      selected = false;
    }
  }
  return selected;
};

export const useHandleBrokenPage = () => {
  const [broken, setBroken] = useState<boolean>(false);
  const [errorMessage, setErrorMessage] = useState<string | undefined>();

  /** Convenience function to handle cases where we must abort and tell the user we have done so */
  const handleBrokenPage = (message: string) => {
    displayError(Error(message));
    setErrorMessage(message);
    setBroken(true);
  };

  return { broken, errorMessage, handleBrokenPage };
};
