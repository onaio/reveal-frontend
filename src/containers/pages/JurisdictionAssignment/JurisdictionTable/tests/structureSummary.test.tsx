import { Dictionary } from '@onaio/utils';
import { mount, ReactWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter, Route, RouteComponentProps, Switch } from 'react-router';
import { AUTO_ASSIGN_JURISDICTIONS_URL } from '../../../../../constants';
import store from '../../../../../store';
import { fetchTree, selectNode } from '../../../../../store/ducks/opensrp/hierarchies';
import { raZambiaHierarchy } from '../../../../../store/ducks/opensrp/hierarchies/tests/fixtures';
import { irsPlans } from '../../../../../store/ducks/opensrp/PlanDefinition/tests/fixtures';
import { ConnectedSelectedStructuresTable } from '../structureSummary';

/** will use this to render table rows */
const renderTable = (wrapper: ReactWrapper, message: string) => {
  const trs = wrapper.find('table tr');
  trs.forEach(tr => {
    expect(tr.text()).toMatchSnapshot(message);
  });
};

describe('', () => {
  /** nothing but if drill down works, ooh and the structure total count is correct */
  it('structure Summary works correctly ', () => {
    const plan = irsPlans[0];
    // prepare store: should have a tree with  at-least one selected leaf node
    store.dispatch(fetchTree(raZambiaHierarchy));
    // select Akros_1   3951
    store.dispatch(selectNode('0ddd9ad1-452b-4825-a92a-49cb9fc82d18', '3951'));

    /** current architecture does not use the Jurisdiction table as a view
     * it is seen as a controlled component that is feed some data from controlling component
     * That's why there was a need to have this custom view and a mock routing system.
     */
    const CustomView = (props: RouteComponentProps<Dictionary>) => {
      const mockProps = {
        autoSelectionFlow: false,
        currentParentId: props.match.params.parentId,
        rootJurisdictionId: '0ddd9ad1-452b-4825-a92a-49cb9fc82d18',
      };
      return <ConnectedSelectedStructuresTable {...mockProps} />;
    };

    const App = () => {
      return (
        <Switch>
          <Route
            exact={true}
            path={`${AUTO_ASSIGN_JURISDICTIONS_URL}/:planId/:rootId/:parentId`}
            component={CustomView}
          />
          <Route
            exact={true}
            path={`${AUTO_ASSIGN_JURISDICTIONS_URL}/:planId`}
            component={CustomView}
          />
        </Switch>
      );
    };

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[
            {
              hash: '',
              pathname: `${AUTO_ASSIGN_JURISDICTIONS_URL}/${plan.identifier}`,
              search: '',
              state: {},
            },
          ]}
        >
          <App />
        </MemoryRouter>
      </Provider>
    );

    renderTable(wrapper, 'should have single node(parent node)');
    const tbodyRow = wrapper.find('tbody tr');
    expect(tbodyRow.length).toEqual(1);
    expect(
      tbodyRow
        .at(0)
        .text()
        .includes('Zambia')
    ).toBeTruthy();

    // simulate click on checkbox to check
    // tbodyRow.find('input').simulate('change', { target: { name: '', checked: true } });
    // wrapper.update();
    // renderTable(wrapper, 'input checked');

    expect(
      wrapper
        .find('tbody tr')
        .at(0)
        .text()
    ).toMatchInlineSnapshot(`"ra Zambia13"`);

    // test drilldown
    tbodyRow
      .at(0)
      .find('NodeCell Link a')
      .simulate('click', { button: 0 });

    wrapper.update();
    renderTable(wrapper, 'after first click');

    expect(
      wrapper
        .find('tbody tr')
        .at(0)
        .text()
    ).toMatchInlineSnapshot(`"Lusaka13"`);

    wrapper
      .find('tbody tr')
      .at(0)
      .find('NodeCell a')
      .simulate('click', { button: 0 });

    wrapper.update();
    renderTable(wrapper, 'after second click');

    expect(
      wrapper
        .find('tbody tr')
        .at(0)
        .text()
    ).toMatchInlineSnapshot(`"Mtendere13"`);

    wrapper
      .find('tbody tr')
      .at(0)
      .find('NodeCell a')
      .simulate('click', { button: 0 });

    wrapper.update();
    renderTable(wrapper, 'after third click');

    expect(
      wrapper
        .find('tbody tr')
        .at(0)
        .text()
    ).toMatchInlineSnapshot(`"Akros_113"`);

    // akros is the last child it should not be clickable
    expect(toJson(wrapper.find('tbody tr NodeCell span'))).toMatchSnapshot(
      'should not have btn-link or onClick handler'
    );
  });
});
