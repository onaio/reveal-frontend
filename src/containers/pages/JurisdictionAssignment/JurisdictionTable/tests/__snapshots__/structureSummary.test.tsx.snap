// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`structure Summary works correctly : after first click 1`] = `"NameStructures CountSelected Jurisdictions"`;

exports[`structure Summary works correctly : after first click 2`] = `"Lusaka13"`;

exports[`structure Summary works correctly : after second click 1`] = `"NameStructures CountSelected Jurisdictions"`;

exports[`structure Summary works correctly : after second click 2`] = `"Mtendere13"`;

exports[`structure Summary works correctly : after third click 1`] = `"NameStructures CountSelected Jurisdictions"`;

exports[`structure Summary works correctly : after third click 2`] = `"Akros_113"`;

exports[`structure Summary works correctly : should have single node(parent node) 1`] = `"NameStructures CountSelected Jurisdictions"`;

exports[`structure Summary works correctly : should have single node(parent node) 2`] = `"ra Zambia13"`;

exports[`structure Summary works correctly : should not have btn-link or onClick handler 1`] = `
<span
  key="3951-span"
>
  Akros_1
</span>
`;
